### ISSUE TITLE
Padding Mask部分的疑惑

### ISSUE THREAD

**Issue by skygreygrey** [jingyaogong/minimind#81]
您好，感谢您的项目。
我在您的项目中发现模型中的Attention模块中的mask只是包含 attention mask，而没有通过传入padding部分的padding mask结合作为mask。我发现您在模型输出完logits后计算loss时（例如pretrain代码中计算entropy loss），通过padding mask对loss计算结果mask掉了padding部分。我想请教下这两种padding mask的实现方法上在效果上有什么区别。
谢谢，希望您能回复我的疑惑！

**Comment by jingyaogong** [comment]
## 1、Attention Mask的作用
假设模型需要学习一句话："我来自地球<终止符>"

"我来自地球" 输入一次，模型只能够学习一次，也就是学会预测最后的 "<终止符>"
然而，这样太慢了，现在希望它只需要接收1次输入，就能等价于学5次。

自然可以想到将输入的 “我来自地球” 展开成一个矩阵，就可以愉快地并行训练了，每一行被学习的时候，我们的目的是：

```
* 希望它看到 “我” 能学会预测 “来” 
* 希望它看到 “我来” 能学会预测 “自”
* 希望它看到 “我来自” 能学会预测“地”
* .......
```

但是完全展开成矩阵：
```
我来自地球
我来自地球
我来自地球
我来自地球
我来自地球
```
在第 i 行，无论 i 取几，模型都可以看到一整行的内容，下一个要预测啥，下下一个要预测啥，答案都一眼看到了，显然答案被开卷就没有意义了。

因此这里需要把每一行不该看到的内容给加上Mask矩阵，使之变成负无穷（相当于注意力分数变成0，把这个位置的token给藏起来即可）：
```
我负负负负
我来负负负
我来自负负
我来自地负
我来自地球
```
这样的话，每行各司其职，预测它该预测的值

```
* 第一行学习“来” (模型只参考"我")
* 第二行学习“自” (模型只参考"我来")
* 第三行学习“地” (模型只参考"我来自")
* 第四行学习“球” (模型只参考"我来自地")
* 第五行学习“<终止符>” (模型只参考"我来自地球")
```

完全做到各司其职，互不干扰，过去的token无法看见未来的token。

这里的Attention Mask达到**只输入一次，实际上潜在地并行训练了N次的目的**

## 2、logit的mask

这里没什么特殊的含义，如果设定训练长度为512。

但是一段话只有200个有效token需要被学习，那么相当于需要padding 312个无实际含义的token使之统一长度方便输入训练。

那么模型的推理过程并不需要关注这312个无意义的token，所以直接在最后计算loss时去除这部分的损失即可。

至于Attention中无论是否计算这312个token都没有关系，因为输出的loss反正都会屏蔽掉这部分损失，所以Attention模块只需要面无表情地麻木工作即可，它并不需要知道token数量到底是512还是312还是200。因为这部分工作由loss完成。

举个不恰当的例子，

高考数学时，监考老师并不会要求学生格式化大脑中的 **语英政史地生物化** 知识只集中Attention Mask矩阵计算数学题。
因为这张试卷高考Loss只体现在头脑风暴的数学答案中。

**Comment by skygreygrey** [comment]
感谢您的回复。

**Comment by cpp2016** [comment]
至于Attention中无论是否计算这312个token都没有关系，因为输出的loss反正都会屏蔽掉这部分损失，所以Attention模块只需要面无表情地麻木工作即可，它并不需要知道token数量到底是512还是312还是200。因为这部分工作由loss完成。

这句话中，后边312个token可能会对前边200个token产生贡献吧，特别是在上层，注意力中是不是应该屏蔽掉

**Comment by jingyaogong** [comment]
> 至于Attention中无论是否计算这312个token都没有关系，因为输出的loss反正都会屏蔽掉这部分损失，所以Attention模块只需要面无表情地麻木工作即可，它并不需要知道token数量到底是512还是312还是200。因为这部分工作由loss完成。
> 
> 这句话中，后边312个token可能会对前边200个token产生贡献吧，特别是在上层，注意力中是不是应该屏蔽掉

不会，GPT属于Causal Model，前N个位置的Attention计算不会考虑N+1及以后的未来token，只会关注当前位置之前的token。

相反，如果312个空token padding在前面而不是后面，会对剩下200个原有token产生影响。
